File: pid.c
struct PIDConfig {
  const long p;
  const long i;
  const long d;
  long last_pos;
  long integ;
};

void Init(struct PIDConfig *config) {
  config->last_pos = 0;
  config->integ = 0;
}

long Calculate(struct PIDConfig *config, long pos, long goal) {
  long error = goal - pos;
  long deriv = pos - config->last_pos;
  config->integ += error;
  config->last_pos = pos;
  return (config->p * error) - (config->d * deriv) + (config->i * config->integ);
}

File: pid.rs
trait PIDController {
    fn calculate(&mut self, position: f32, goal: f32) -> f32;
}

struct PIDConfig {
    p: f32,
    i: f32,
    d: f32,
    last_position: f32,
    integral: f32,
}

impl PIDController for PIDConfig {
    fn calculate(&mut self, position: f32, goal: f32) -> f32 {
        let error = goal - position;
        let deriv = self.last_position - position;
        self.last_position = position;
        self.integral += error;
        return (self.p * error) + (self.i * self.integral) - (self.d * deriv);
    }
}
